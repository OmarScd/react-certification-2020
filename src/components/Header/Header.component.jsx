import React from 'react';
import { Link } from 'react-router-dom';
import styled from 'styled-components';

const Container = styled.header`
  background: ${(props) => props.theme.colors.primary};
  display: flex;
  justify-content: space-between;
  padding: 1rem 2rem;
  width: 100%;
`;

const HeaderSection = styled.div`
  display: flex;
`;

const HeaderElement = styled.div`
  margin: 0;
  padding: 0 1rem;
`;

const NavLink = styled(Link)`
  color: white;
  font-weight: bold;
  text-decoration: none;

  :hover {
    color: #dedede;
    transition: color 0.3s ease-in-out;
  }
`;

const Header = () => {
  return (
    <Container>
      <HeaderSection>
        <HeaderElement>
          <NavLink to="/fake-route">Test</NavLink>
        </HeaderElement>
        <HeaderElement>2</HeaderElement>
      </HeaderSection>
      <HeaderSection>
        <HeaderElement>3</HeaderElement>
        <HeaderElement>4</HeaderElement>
      </HeaderSection>
    </Container>
  );
};

export default Header;
