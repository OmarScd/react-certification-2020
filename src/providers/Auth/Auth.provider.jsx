import React, { useState, useEffect, useContext } from 'react';

import { USERNAME_STORAGE_KEY } from '../../utils/constants';
import { storage } from '../../utils/storage';

const AuthContext = React.createContext(null);

const useAuth = () => {
  const context = useContext(AuthContext);
  if (!context) {
    throw new Error(`Can't use "useAuth" without an AuthProvider!`);
  }
  return context;
};

const AuthProvider = ({ children }) => {
  const [authenticated, setAuthenticated] = useState(false);
  const [username, setUsername] = useState(undefined);

  useEffect(() => {
    const user = storage.get(USERNAME_STORAGE_KEY);
    const isAuthenticated = !!user;

    setAuthenticated(isAuthenticated);
    if (isAuthenticated) {
      setUsername(user);
    }
  }, []);

  const login = (name) => {
    setAuthenticated(true);
    setUsername(name);
    storage.set(USERNAME_STORAGE_KEY, name);
  };

  const logout = () => {
    setAuthenticated(false);
    setUsername(undefined);
    storage.remove(USERNAME_STORAGE_KEY);
  };

  return (
    <AuthContext.Provider value={{ login, logout, authenticated, username }}>
      {children}
    </AuthContext.Provider>
  );
};

export { useAuth, AuthProvider as default };
