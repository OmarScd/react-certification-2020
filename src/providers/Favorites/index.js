export { default } from './Favorites.provider';
