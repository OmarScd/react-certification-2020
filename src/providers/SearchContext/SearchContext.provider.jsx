import React, { useState, useContext } from 'react';

const SearchContext = React.createContext();

const useSearch = () => {
  const context = useContext(SearchContext);
  if (!context) {
    throw new Error(`Can't use "useSearch" without a FavoritesProvider!`);
  }
  return context;
};

const SearchContextProvider = ({ children }) => {
  const [query, setQuery] = useState('Subdivisions');

  return (
    <SearchContext.Provider value={{ query, setQuery }}>
      {children}
    </SearchContext.Provider>
  );
};

export { useSearch, SearchContextProvider as default };
