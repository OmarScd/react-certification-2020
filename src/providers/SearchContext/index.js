export { default, useSearch } from './SearchContext.provider';
