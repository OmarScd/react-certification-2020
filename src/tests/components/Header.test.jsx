import React from 'react';
import { render, screen, cleanup, fireEvent } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import Header from '../../components/Header';
import ThemeProvider from '../../providers/Theme';
import '../../providers/SearchContext';
import '../../providers/Auth';

jest.mock('../../providers/SearchContext', () => ({
  default: jest.fn(),
  useSearch: () => ({ query: 'test', setQuery: jest.fn() }),
}));

jest.mock('../../providers/Auth', () => ({
  default: jest.fn(),
  useAuth: () => {
    let authenticated = true;
    return {
      logout: jest.fn(() => {
        authenticated = false;
      }),
      authenticated,
    };
  },
}));

describe('Header logged in', () => {
  afterAll(cleanup);

  it('renders correctly when logged in', () => {
    const { container } = render(
      <ThemeProvider>
        <BrowserRouter>
          <Header />
        </BrowserRouter>
      </ThemeProvider>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders correctly when logged out', () => {
    const { container } = render(
      <ThemeProvider>
        <BrowserRouter>
          <Header />
        </BrowserRouter>
      </ThemeProvider>
    );
    const logOutButton = screen.getByRole('button');
    fireEvent.click(logOutButton);
    expect(container.firstChild).toMatchSnapshot();
  });
});
