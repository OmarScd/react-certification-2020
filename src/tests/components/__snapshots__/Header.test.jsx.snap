// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header logged in renders correctly when logged in 1`] = `
<header
  class="sc-hKFyIo jcdTAl"
>
  <div
    class="sc-eCApGN hCdAiT"
  >
    <div
      class="sc-jSFkmK qmXvl"
    >
      <a
        class="sc-gKAblj iUhdyg"
        href="/"
      >
        Home
      </a>
    </div>
    <div
      class="sc-jSFkmK qmXvl"
    >
      <a
        class="sc-gKAblj iUhdyg"
        href="/favorites"
      >
        Favorites
      </a>
    </div>
    <div
      class="sc-jSFkmK qmXvl"
    >
      <form
        class="sc-bdnylx hYwFxl"
      >
        <img
          alt="magnifying lens"
          class="sc-dlnjPT dGgzTM"
          src="search-24px.svg"
        />
        <input
          class="sc-gtssRu fkigkU"
          placeholder="Search..."
          type="text"
          value="test"
        />
      </form>
    </div>
  </div>
  <div
    class="sc-eCApGN hCdAiT"
  >
    <div
      class="sc-jSFkmK qmXvl"
    >
      <button
        class="sc-iCoHVE kHdmrf"
      >
        Log out
      </button>
    </div>
    <div
      class="sc-jSFkmK qmXvl"
    >
      <img
        alt="person"
        class="sc-fujyUd hfIOzZ"
        src="person-24px.svg"
      />
    </div>
  </div>
</header>
`;

exports[`Header logged in renders correctly when logged out 1`] = `
<header
  class="sc-hKFyIo jcdTAl"
>
  <div
    class="sc-eCApGN hCdAiT"
  >
    <div
      class="sc-jSFkmK qmXvl"
    >
      <a
        class="sc-gKAblj iUhdyg"
        href="/"
      >
        Home
      </a>
    </div>
    <div
      class="sc-jSFkmK qmXvl"
    >
      <a
        class="sc-gKAblj iUhdyg"
        href="/favorites"
      >
        Favorites
      </a>
    </div>
    <div
      class="sc-jSFkmK qmXvl"
    >
      <form
        class="sc-bdnylx hYwFxl"
      >
        <img
          alt="magnifying lens"
          class="sc-dlnjPT dGgzTM"
          src="search-24px.svg"
        />
        <input
          class="sc-gtssRu fkigkU"
          placeholder="Search..."
          type="text"
          value="test"
        />
      </form>
    </div>
  </div>
  <div
    class="sc-eCApGN hCdAiT"
  >
    <div
      class="sc-jSFkmK qmXvl"
    >
      <button
        class="sc-iCoHVE kHdmrf"
      >
        Log out
      </button>
    </div>
    <div
      class="sc-jSFkmK qmXvl"
    >
      <img
        alt="person"
        class="sc-fujyUd hfIOzZ"
        src="person-24px.svg"
      />
    </div>
  </div>
</header>
`;
